### PR TITLE
fix: check pkg main

### PR DESCRIPTION
**Pull request by calebboyd**
This correctly identifies the main file (afaict)

But its still not resolving to the correct module. (null at runtime), so some tests are probably broken

### CODE REVIEW

**Comment by nchanged** [comment]
Seems like a legit fix to me, let's figure out what's wrong with tests, we need to make them pass

**Comment by calebboyd** [comment]
linking: #795

**Comment by calebboyd** [comment]
So this fix doesn't work.. It works in the context of "`testFolder`" But the bundle is not resolving the correct file. If you can point me in the right direction I can continue to investigate, but Im not very familiar with this portion of fusebox

**Comment by nchanged** [comment]
well it should be here
https://github.com/fuse-box/fuse-box/blob/91b4c47843eb11b4521af202f8875d1b6accfa65/src/core/PathMaster.ts#L233-L246

That's for sure, and has to return alias:

https://github.com/fuse-box/fuse-box/blob/91b4c47843eb11b4521af202f8875d1b6accfa65/src/core/PathMaster.ts#L296-L299

Actually, alias should come here:

https://github.com/fuse-box/fuse-box/blob/91b4c47843eb11b4521af202f8875d1b6accfa65/src/core/PathMaster.ts#L311-L313

**Comment by calebboyd** [comment]
I'm out for the night. I think I've fixed the issue. Please review!

File: gulpfile.js [hunk 1]
```diff
@@ -326,12 +326,8 @@ gulp.task("installDevDeps", function(done) {
         "buble",
     ];
 
-    if (os.platform().match(/^win/)) {
-        let windowsCommands = ["start", "cmd.exe", "/K", "npm", "install", "--no-save", ...deps];
```
Comment (nchanged) [inline gulpfile.js]: why did you remove that one?

File: src/loader/LoaderAPI.ts [hunk 2]
```diff
@@ -293,9 +293,14 @@ function $getRef(name: string, o: RefOpts): IReference {
         wildcard = validPath;
     }
     if (!file && !wildcard) {
-        // try folder index.js
+        // try index.js
         validPath = $pathJoin(filePath, "/", "index.js");
         file = pkg.f[validPath];
+
+        if (!file && pkg.s && pkg.s.entry) {
```
Comment (nchanged) [inline src/loader/LoaderAPI.ts]: what's that? Why did you modify the loader?

File: gulpfile.js (hunk 1 above)
Comment (calebboyd) [reply gulpfile.js]: I replaced it with `spawn('npm.cmd', ...)` on windows

This avoids a shell popping up, that has to be closed manually

File: src/loader/LoaderAPI.ts (hunk 2 above)
Comment (calebboyd) [reply src/loader/LoaderAPI.ts]: If the path is fully relative it will land here. See:
https://github.com/calebboyd/fuse-test 

If its not the index it must be an entry file. Specifically, a fully relative name.

`'..'` or `'../../'`. As a guard an `isRelative` check could be made..

Thoughts?

Comment (nchanged) [reply src/loader/LoaderAPI.ts]: But what if there a package.json with a different entry? that's why we had that `alias` thingy in PathMaster, that overrides it

Comment (calebboyd) [reply src/loader/LoaderAPI.ts]: If you have an idea of what `alias` needs to be set to, any tips would be helpful.

Comment (calebboyd) [reply src/loader/LoaderAPI.ts]: From what I can tell providing an alias does nothing -- The property is set but then it is never  used again...
Is that correct?
https://github.com/fuse-box/fuse-box/blob/813bc793c7d9de764a5f9aac64889233426f2c0a/src/core/PathMaster.ts#L153-L155

File: src/core/PathMaster.ts [hunk 3]
```diff
@@ -328,7 +328,7 @@ export class PathMaster {
                 let folder = path.isAbsolute(name) ? name : path.join(root, name);
                 const folderPath = this.testFolder(folder, name);
                 if (folderPath) {
-                    return { resolved: folderPath }
+                    return { resolved: folderPath, alias: name }
```
Comment (calebboyd) [inline src/core/PathMaster.ts]: Adding alias here doesn't really do anything

File: src/core/PathMaster.ts [hunk 4]
```diff
@@ -224,33 +226,49 @@ export class PathMaster {
...
-                    let result = path.join(name, index);
-                    let startsWithDot = result[0] === "."; // After transformation we need to bring the dot back
-                    if (startsWithDot) {
-                        result = `./${result}`;
-                    }
-                    return result;
+            let result
+            const pkg = this.tryReadPkg(folder)
+            if (pkg && pkg.main) {
+                result = path.join(name, pkg.main)
```
Comment (nchanged) [inline src/core/PathMaster.ts]: This should return an alias (FuseBox will regenerate the code and point to a correct file)

File: src/loader/LoaderAPI.ts [hunk 5]
```diff
@@ -293,9 +293,13 @@ function $getRef(name: string, o: RefOpts): IReference {
         wildcard = validPath;
     }
     if (!file && !wildcard) {
-        // try folder index.js
+        // try index.js
         validPath = $pathJoin(filePath, "/", "index.js");
         file = pkg.f[validPath];
+
+        if (!file && filePath === ".") {
```
Comment (nchanged) [inline src/loader/LoaderAPI.ts]: Do we still need to modify the loader? Which case does it cover? `require(".")`?
Shouldn't it require index if exists?

**Comment by nchanged** [comment]
Amazing, almost there, let's verify a few things before merging

File: src/loader/LoaderAPI.ts (hunk 5 above)
Comment (calebboyd) [reply src/loader/LoaderAPI.ts]: Yeah, So here is an example of it being hit
https://github.com/calebboyd/fuse-test

Adding a log in that if block (L301)
`console.log('loading: "', name, '" filePath: "', filePath, '" entry: "', pkg.s.entry, '"')`

-> 
```
loading: " ../ " filePath: " . " entry: " test.js "
loading: " ../../ " filePath: " . " entry: " test.js "
```

Comment (calebboyd) [reply src/loader/LoaderAPI.ts]: It covers any fully relative path that resolves to the package.json in the pkg context

**Comment by dgreif** [comment]
Bundled a new project with fusebox today.  It uses `request-promise`, which does some interesting require paths in `require-promise-core`.  It failed with `2.2.31`, but worked with `2.3.1-beta.14`.  This is a real-world case where @calebboyd's fix is needed.
